File: frontend/src/types/market.ts
/**
 * 行情域契约（总览页所需子集）：情绪指标与历史。
 *
 * 逐字段镜像后端 `app/schemas/market.py` / `app/schemas/common.py`，
 * 字段名保持 snake_case。比率字段为**小数口径**（`0.42 = 42%`）。
 *
 * **注意**：本文件**不**从 `types/index.ts` 导出（该出口由其他 Task 维护），
 * 使用方一律 `import type { ... } from '@/types/market'`。
 */

/** 行情数据新鲜度标记（后端 `MarketMeta`）。 */
export interface MarketMeta {
  /** 数据是否已超出预期新鲜度窗口（上游不可用时的降级返回） */
  stale: boolean;
  /** 库中最新数据日期 */
  data_date: string | null;
}

/** 市场情绪指标。 */
export interface SentimentOut {
  trade_date: string;
  /** 情绪温度（0~100 分） */
  temperature: number;
  /** 情绪周期阶段（冰点 / 冰点转折 / 修复 / 加速/高潮 / 分歧 / 退潮）；未派生时为 null */
  stage: string | null;
  limit_up_count: number;
  limit_down_count: number;
  broken_board_count: number;
  /** 炸板率（小数口径） */
  broken_rate: number;
  up_count: number;
  down_count: number;
  max_continue_days: number | null;
  /** 昨日涨停今日平均溢价（小数口径） */
  premium_rate: number;
}

/** 某交易日情绪指标。 */
export interface SentimentResponse extends MarketMeta {
  trade_date: string | null;
  item: SentimentOut | null;
}

/** 最近 N 个交易日情绪（按交易日升序）。 */
export interface SentimentHistoryResponse extends MarketMeta {
  days: number;
  items: SentimentOut[];
}

// --------------------------------------------------------------- 通用外壳

/** 分页外壳字段（后端 `PageResponse` / `MarketPageResponse` 共有部分）。 */
export interface PageMeta {
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

/** 可选交易日列表。 */
export interface DatesResponse {
  dates: string[];
  limit: number;
}

// --------------------------------------------------------------- 涨停池

/** 关联板块（涨停原因所属题材）。 */
export interface PoolPlate {
  plate_id?: number | null;
  plate_name?: string | null;
}

/** 封板时间线单点（status: 1 封涨停 / 2 炸板 / 3 封跌停 / 4 开跌停）。 */
export interface PoolTimelinePoint {
  /** Unix 秒 */
  timestamp: number;
  status: number;
}

/** 涨停池成分（`turnover_rate` / `change_pct` / `seal_ratio` 小数口径）。 */
export interface PoolOut {
  trade_date: string;
  code: string;
  name: string;
  continue_days: number;
  limit_up_time: string | null;
  seal_amount_yuan: number | null;
  open_times: number | null;
  turnover_rate: number | null;
  amount_yuan: number | null;
  /** 总市值（元） */
  market_cap_yuan: number | null;
  /** 现价（元） */
  price: number | null;
  /** 涨跌幅（小数，0.1001 = +10.01%） */
  change_pct: number | null;
  /** 量比 */
  volume_bias_ratio: number | null;
  /** 流通市值（元） */
  free_cap_yuan: number | null;
  /** 封单比（小数） */
  seal_ratio: number | null;
  /** 涨停原因（上游口语化说明） */
  reason: string | null;
  /** 关联板块 */
  plates: PoolPlate[] | null;
  /** 封板时间线 */
  timeline: PoolTimelinePoint[] | null;
  /** 池型（对齐上游 pool_name，共 7 种） */
  pool_type: string;
}

/** 某交易日全部池型。 */
export interface PoolsResponse extends MarketMeta {
  trade_date: string | null;
  pools: Record<string, PoolOut[]>;
}

/** 某交易日单个池型成分。 */
export interface PoolResponse extends MarketMeta {
  trade_date: string | null;
  pool_type: string;
  min_continue_days: number;
  items: PoolOut[];
}

// --------------------------------------------------------------- 连板天梯

/** 天梯单元格：某票某交易日的连板情况。 */
export interface LadderCell {
  /** 该日连板天数 */
  boards: number;
  /** 该日首封时间 HH:MM */
  first_seal_time: string | null;
}

/** 天梯矩阵的一行（一只个股，跨多日的连板单元格）。 */
export interface LadderMatrixRow {
  code: string;
  name: string;
  /** 交易日（ISO 字符串）→ 该日单元格 */
  cells: Record<string, LadderCell>;
}

/** 连板天梯矩阵：列=交易日（升序），行=个股。 */
export interface LadderMatrixResponse extends MarketMeta {
  /** 交易日（ISO 字符串，升序） */
  days: string[];
  rows: LadderMatrixRow[];
  min_continue_days: number;
}

/** 连板天梯单行。 */
export interface LadderRowOut {
  trade_date: string;
  code: string;
  name: string;
  continue_days: number;
  first_seal_time: string | null;
}

/** 连板天梯分页响应。 */
export interface LadderPage extends PageMeta, MarketMeta {
  items: LadderRowOut[];
}

// --------------------------------------------------------------- 快讯

/** 7×24 快讯。 */
export interface NewsFlashOut {
  ts: string;
  level: string | null;
  title: string;
  summary: string | null;
  symbols: string[];
  categories: string[];
}

/** 快讯分页响应。 */
export interface NewsFlashPage extends PageMeta, MarketMeta {
  items: NewsFlashOut[];
}

// --------------------------------------------------------------- 主题

/** 主题强度榜单项（`core_avg_pct` 小数口径）。 */
export interface ThemeOut {
  trade_date: string;
  rank: number;
  name: string;
  core_avg_pct: number | null;
  description: string | null;
  core_count: number | null;
}

/** 某交易日主题强度榜。 */
export interface ThemesResponse extends MarketMeta {
  trade_date: string | null;
  items: ThemeOut[];
}

/** 主题成分股（`pct` / `turnover_rate` 小数口径）。 */
export interface ThemeStockOut {
  trade_date: string;
  theme_name: string;
  code: string;
  name: string;
  price: number;
  pct: number;
  turnover_rate: number;
  continue_days: number | null;
  selected_at: string | null;
}

/** 某日某主题的成分股。 */
export interface ThemeStocksResponse extends MarketMeta {
  trade_date: string;
  theme_name: string;
  items: ThemeStockOut[];
}

// --------------------------------------------------------------- 监管名单

/** 监管名单记录。 */
export interface MonitorStockOut {
  trade_date: string;
  kind: string;
  code: string;
  name: string;
  reason: string | null;
}

/** 某日监管名单。 */
export interface MonitorResponse extends MarketMeta {
  trade_date: string | null;
  kind: string | null;
  items: MonitorStockOut[];
}

File: backend/app/schemas/market.py
"""行情域响应契约：股票、日线/分时、涨停池、天梯、情绪、主题、快讯、监管名单。

字段名与单位沿用 ``app/models/market.py`` 的领域标准名（价格=元，量=股/手，
比率=小数口径）。行情类响应继承 :class:`~app.schemas.common.MarketMeta`，携带
``stale`` / ``data_date`` 新鲜度标记。
"""

from __future__ import annotations

from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import MarketMeta

__all__ = [
    "DailyBarOut",
    "DailyBarsResponse",
    "LadderRowOut",
    "LimitUpPoolOut",
    "MinuteBarOut",
    "MinuteBarsResponse",
    "MonitorResponse",
    "MonitorStockOut",
    "NewsFlashOut",
    "PoolResponse",
    "PoolsResponse",
    "SentimentHistoryResponse",
    "SentimentOut",
    "SentimentResponse",
    "StockOut",
    "ThemeOut",
    "ThemeStockOut",
    "ThemeStocksResponse",
    "ThemesResponse",
]

_ORM = ConfigDict(from_attributes=True)


class StockOut(BaseModel):
    """股票基础信息。"""

    model_config = _ORM

    code: str
    name: str
    market: str
    board: str
    is_st: bool
    list_date: date | None = None


class DailyBarOut(BaseModel):
    """日线行情（价格元，``volume_shares`` 股，``amount_yuan`` 元）。"""

    model_config = _ORM

    code: str
    trade_date: date
    open: float
    high: float
    low: float
    close: float
    pre_close: float | None = None
    volume_shares: int
    amount_yuan: float


class DailyBarsResponse(MarketMeta):
    """某只股票区间日线序列。"""

    code: str
    start: date
    end: date
    items: list[DailyBarOut] = Field(default_factory=list)


class MinuteBarOut(BaseModel):
    """分时行情（``price`` 元，``volume_lots`` 手；``amount_yuan`` eltdx 源为 null）。"""

    model_config = _ORM

    code: str
    trade_date: date
    minute_index: int
    time_label: str
    price: float
    volume_lots: int
    amount_yuan: float | None = None


class MinuteBarsResponse(MarketMeta):
    """某只股票某日分时序列。"""

    code: str
    trade_date: date
    items: list[MinuteBarOut] = Field(default_factory=list)


class LimitUpPoolOut(BaseModel):
    """涨停池成分（``turnover_rate`` / ``change_pct`` / ``seal_ratio`` 小数口径）。"""

    model_config = _ORM

    trade_date: date
    code: str
    name: str
    continue_days: int
    limit_up_time: str | None = None
    seal_amount_yuan: float | None = None
    open_times: int | None = None
    turnover_rate: float | None = None
    amount_yuan: float | None = None
    market_cap_yuan: float | None = None
    price: float | None = None
    change_pct: float | None = None
    volume_bias_ratio: float | None = None
    free_cap_yuan: float | None = None
    seal_ratio: float | None = None
    reason: str | None = None
    plates: list[dict[str, Any]] | None = None
    timeline: list[dict[str, Any]] | None = None
    pool_type: str


class PoolsResponse(MarketMeta):
    """某交易日全部池型（``pool_type -> 成分``）。"""

    trade_date: date | None = None
    pools: dict[str, list[LimitUpPoolOut]] = Field(default_factory=dict)


class PoolResponse(MarketMeta):
    """某交易日单个池型成分。"""

    trade_date: date | None = None
    pool_type: str
    min_continue_days: int
    items: list[LimitUpPoolOut] = Field(default_factory=list)


class LadderRowOut(BaseModel):
    """连板天梯单行。"""

    model_config = _ORM

    trade_date: date
    code: str
    name: str
    continue_days: int
    first_seal_time: str | None = None


class SentimentOut(BaseModel):
    """市场情绪指标（``temperature`` 分，比率字段小数口径）。"""

    model_config = _ORM

    trade_date: date
    temperature: float
    stage: str | None = None
    limit_up_count: int
    limit_down_count: int
    broken_board_count: int
    broken_rate: float
    up_count: int
    down_count: int
    max_continue_days: int | None = None
    premium_rate: float


class SentimentResponse(MarketMeta):
    """某交易日情绪指标。"""

    trade_date: date | None = None
    item: SentimentOut | None = None


class SentimentHistoryResponse(MarketMeta):
    """最近 ``days`` 个交易日情绪（按交易日升序）。"""

    days: int
    items: list[SentimentOut] = Field(default_factory=list)


class ThemeOut(BaseModel):
    """主题强度榜单项。"""

    model_config = _ORM

    trade_date: date
    rank: int
    name: str
    core_avg_pct: float | None = None
    description: str | None = None
    core_count: int | None = None


class ThemesResponse(MarketMeta):
    """某交易日主题强度榜。"""

    trade_date: date | None = None
    items: list[ThemeOut] = Field(default_factory=list)


class ThemeStockOut(BaseModel):
    """主题成分股。"""

    model_config = _ORM

    trade_date: date
    theme_name: str
    code: str
    name: str
    price: float
    pct: float
    turnover_rate: float
    continue_days: int | None = None
    selected_at: datetime | None = None


class ThemeStocksResponse(MarketMeta):
    """某日某主题的成分股。"""

    trade_date: date
    theme_name: str
    items: list[ThemeStockOut] = Field(default_factory=list)


class NewsFlashOut(BaseModel):
    """快讯。"""

    model_config = _ORM

    ts: datetime
    level: str | None = None
    title: str
    summary: str | None = None
    symbols: list[str] = Field(default_factory=list)
    categories: list[str] = Field(default_factory=list)


class MonitorStockOut(BaseModel):
    """监管名单记录。"""

    model_config = _ORM

    trade_date: date
    kind: str
    code: str
    name: str
    reason: str | None = None


class MonitorResponse(MarketMeta):
    """某日监管名单。"""

    trade_date: date | None = None
    kind: str | None = None
    items: list[MonitorStockOut] = Field(default_factory=list)
